### PR TITLE
 PFM-ISSUE-10314: Generate package.json file for all plugins with assets 

### PR DESCRIPTION
**Pull request by vladimir-arsov**
Resolves [PFM-ISSUE-10314](https://base.cplace.io/pages/hdp18naubqph9koaumm7a9rlb/PFM-ISSUE-10314-Generate-package.json-file-for-all-plugins-with-assets)

Depends on [PFM-ISSUE-10350](https://github.com/collaborationFactory/cplace-asc/pull/80).

### CODE REVIEW

File: src/model/CplacePlugin.ts [hunk 1]
```diff
@@ -6,20 +6,56 @@ import * as path from 'path';
...
+    constructor(private plugin: CplacePlugin, repositoryDir: string) {
+        super(repositoryDir);
+    }
+
+    public getFilePath() {
+        return path.join(this.plugin.assetsDir, this.packageJsonFile);
+    }
+
+    public getPackageName() {
+        return `@${this.plugin.pluginDescriptor.group.replace(
```
Comment (slaven3kopic) [inline src/model/CplacePlugin.ts:38]: What is `pluginDescriptor.group`. I've seen only the name so far

File: src/model/CplacePlugin.ts [hunk 2]
```diff
@@ -6,20 +6,56 @@ import * as path from 'path';
...
+    IPackageJsonDependency,
+} from './AbstractPackageJsonGenerator';
+import { isFileTracked } from './utils';
 import { CombineJavascriptsCompiler } from '../compiler/CombineJavascriptsCompiler';
 
 export interface ICplacePluginResolver {
     (pluginName: string): CplacePlugin | undefined;
 }
 
+export class PluginPackageJsonGenerator extends AbstractPackageJsonGenerator {
```
Comment (slaven3kopic) [inline src/model/CplacePlugin.ts:28]: I think this should be moved into a separate file

File: src/model/RootPackageJsonGenerator.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+     *
+     * The dependency plugins are added to the root package.json instead the package.json of each plugin.
+     * That way there won't be duplications of the same dependency in several plugins.
+     */
+    public getPluginDependencies(): IPackageJsonDependency[] {
+        const pluginDependencies: IPackageJsonDependency[] = [];
+
+        this.projects.forEach((value: CplacePlugin, key: string) => {
+            // only go through plugins from current repo
+            if (value.pluginDescriptor.repoName == this.repositoryName) {
```
Comment (slaven3kopic) [inline src/model/RootPackageJsonGenerator.ts:37]: Do we have all of this info in `pluginDescriptor`?

**Review by slaven3kopic — COMMENTED** [review]
Cool

**Comment by slaven3kopic** [comment]
This should be tested pretty good. Maybe with `main` and `cplace-paw` repo together.

File: src/model/RootPackageJsonGenerator.ts (hunk 3 above)
Comment (vladimir-arsov) [reply src/model/RootPackageJsonGenerator.ts:37]: This is available in the new plugin descriptor plugin, starting from 23.1. It includes group id, repo name, version and more details for the dependencies.

**Comment by slaven3kopic** [comment]
@vladimir-arsov For future PRs...Be aware of [this](https://github.com/collaborationFactory/cplace-asc/blob/5ef6b4c843aa2c57e2eb2ea97a87eb1a48b1db94/src/model/CplaceTSConfigGenerator.ts#L38). After parent repos are no longer needed, `tsconfig.base.json` will need to be packaged with the platform and it will need to come from node_modules

**Comment by vladimir-arsov** [comment]
> @vladimir-arsov For future PRs...Be aware of [this](https://github.com/collaborationFactory/cplace-asc/blob/5ef6b4c843aa2c57e2eb2ea97a87eb1a48b1db94/src/model/CplaceTSConfigGenerator.ts#L38). After parent repos are no longer needed, `tsconfig.base.json` will need to be packaged with the platform and it will need to come from node_modules

@slaven3kopic I'll have that in mind. Anyway, I tested this PR for a while now, do you think we can merge it? In any case this feature is activated with a command line parameter so it shouldn't cause any problems.

**Review by slaven3kopic — APPROVED** [review]

File: src/model/CplaceVersion.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            versionParts.push(
+                versionSnapshotParts.length > 1 &&
+                    versionSnapshotParts[1].toLowerCase().includes('snapshot')
+                    ? 'true'
+                    : 'false'
+            );
+
+            if (versionParts.length < 3) {
+                console.error(
+                    cerr`[CplaceVersion] Expected version to consist of 3 parts`
```
Comment (Kaltenbach) [inline src/model/CplaceVersion.ts:64]: For future enhancements: some of our partners use 4 digit versions and other formats. Currently they are definitely not considered to publish artifacts, so this is just to raise the awareness that we might need to adjust this test.

**Review by Kaltenbach — APPROVED** [review]
LGTM

**Pull request merged**